x-pack/plugins/endpoint/public/applications/endpoint/store/index.ts

```diff
@@ -4,25 +4,73 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-import { createStore, compose, applyMiddleware, Store } from 'redux';
+import {
+  createStore,
+  compose,
+  applyMiddleware,
+  Store,
+  Middleware,
+  Dispatch,
+  MiddlewareAPI,
+} from 'redux';
 import { CoreStart } from 'kibana/public';
 import { appSagaFactory } from './saga';
 import { appReducer } from './reducer';
 import { alertMiddlewareFactory } from './alerts/middleware';
+import { routingMiddlewareFactory } from './routing/middleware';
+import { EndpointAppHistory, GlobalState } from '../types';
+import { AppAction } from './action';
 
 const composeWithReduxDevTools = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
   ? (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ name: 'EndpointApp' })
   : compose;
 
-export const appStoreFactory = (coreStart: CoreStart): [Store, () => void] => {
+export type Selector<S, R> = (state: S) => R;
+
+/**
+ * Wrap Redux Middleware and adjust `getState()` to return the namespace from `GlobalState`
+ * that applies to the given Middleware concern.
+ *
+ * @param selector
+ * @param middleware
+ */
+export const substateMiddlewareFactory = <Substate>(
```

Comment:
I think @paul-tavares already has something like this

Reply:
😄 yes... And so does @parkiino . First one to merge wins 💥